config: Enable jh7100 SoC

mcd500 opened the pull request:
Signed-off-by: Akira Tsukamoto <akira.tsukamoto@gmail.com>

farzad64 committed:
nvdla: add NVDLA driver

Additional update from Prashant Gaikwad <pgaikwad@nvidia.com>
Adapted for Linux 5.13 and the BeagleV Starlight board by
<cybergaszcz@gmail.com>

Commit:
sifive/sifive_l2_cache: Add sifive_l2_flush64_range function

Commit:
drivers/dma: Add dw-axi-dmac-starfive driver for VIC7100

Commit:
riscv/starfive: Add VIC7100 support

huanfeng-sf committed:
gpio: starfive-jh7100: Add legacy header and code

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>

huanfeng-sf committed:
drivers/hw_random: Add StarFive JH7100 Random Number Generator driver

Commit:
drivers/pwm: Add SiFive PWM PTC driver

jackzhustf committed:
drivers/video/fbdev and drivers/media/platform: starfive drivers added

1, add ov5640&sc2235 drivers, update stf_isp
2, add MIPI/CSI/DSI drivers for VIC7100

Commit:
riscv: dts: Add JH7100 and BeagleV Starlight support

Based on the device tree in https://github.com/starfive-tech/u-boot/
with contributions from:
yanhong.wang <yanhong.wang@starfivetech.com>
Huan.Feng <huan.feng@starfivetech.com>
ke.zhu <ke.zhu@starfivetech.com>
yiming.li <yiming.li@starfivetech.com>
jack.zhu <jack.zhu@starfivetech.com>
Samin Guo <samin.guo@starfivetech.com>
Chenjieqin <Jessica.Chen@starfivetech.com>
bo.li <bo.li@starfivetech.com>

Rearranged, cleanups, fixes and TPS65086 added by Emil.
Cleanups, fixes, LED and clocks added by Geert.
Cleanups and GPIO fixes from Drew.
Thermal zone added by Stephen.

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>
Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>
Signed-off-by: Stephen L Arnold <nerdboy@gentoo.org>
Signed-off-by: Drew Fustini <drew@beagleboard.org>

Commit:
sifive/sifive_l2_cache: Add disabling IRQ option (workaround)

Commit:
sifive/sifive_l2_cache: Add Starfive support

jackzhustf committed:
video: fbdev: starfive: update fb driver

Commit:
i2c: designware: Add GPIO configuration for JH7100

[FIXME] why we can not do it in U-boot?

[geert: Rebase to v5.13-rc1]
[emil: Initialize gpios before registering driver]

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>

Commit:
drivers/pwm/pwm-sifive-ptc: Clear PWM CNTR

Clear CNTR of PWM after setting period & duty_cycle

jackzhustf committed:
video: fbdev: starfive: workaround for unavailable pointer of dtb

Fix sf_fb_map_video_memory in starfive_fb.c

Commit:
net: stmmac: Configure gtxclk based on speed

esmil committed:
spi: cadence-quadspi: Allow compilation on RISC-V

This IP is also used on the StarFive JH7100 riscv64 SoC and presumably
also the upcoming JH7110 SoC.

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>

mike-scott committed:
drivers/dma: Fix VIC7100 dw-axi-dmac-platform driver addition

Descriptor management was simplified with commit:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=ef6fb2d6f1abd56cc067c694253ea362159b5ac3

Code added to dw-axi-dmac-platform driver due to VIC7100 Cache Coherency
issues needed follow those changes.

Signed-off-by: Michael Scott <mike@foundries.io>

esmil committed:
mfd: tps65086: Make interrupt line optional

The BeagleV Starlight v0.9 board[1] doesn't have the IRQB line routed to
the SoC, but it is still useful to be able to reach the PMIC over I2C
for the other functionality it provides such as GPIOs and regulator
settings.

[1] https://github.com/beagleboard/beaglev-starlight

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>
Acked-for-MFD-by: Lee Jones <lee.jones@linaro.org>

esmil committed:
power: reset: Add TPS65086 restart driver

The only way to reset the BeagleV Starlight v0.9 board[1] properly is to
tell the PMIC to reset itself which will then assert the external reset
lines of the SoC, USB hub and ethernet phy.

This adds a driver to register a reset handler to do just that.

[1] https://github.com/beagleboard/beaglev-starlight

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>
Acked-by: Sebastian Reichel <sebastian.reichel@collabora.com>

esmil committed:
mfd: tps65086: Add cell entry for reset driver

The only way to reset the BeagleV Starlight v0.9 board[1] properly is to
tell the PMIC to reset itself which will then assert the external reset
lines of the SoC, USB hub and ethernet phy.

This adds an mfd cell entry for the reset driver doing just that.

[1] https://github.com/beagleboard/beaglev-starlight

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>

esmil committed:
[NOT-FOR-UPSTREAM] Add build instructions

For convenience this also adds a small beaglev_defconfig and the
firmware needed for the brcmfmac driver along with the signed regulatory
database.

The firmware is from the linux-firmware repo and the regulatory database
from the wireless-regdb Fedora package.

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>
Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>
Signed-off-by: Drew Fustini <drew@beagleboard.org>

esmil committed:
video: fbdev: starfive: fix compiler warnings

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>

esmil committed:
video: fbdev: starfive: improve error handling

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>

mcd500 committed:
config: Enable jh7100 SoC

Signed-off-by: Akira Tsukamoto <akira.tsukamoto@gmail.com>

Commit:
video: fbdev: starfive: fix modpost build error (missing license)

Signed-off-by: Stephen L Arnold <nerdboy@gentoo.org>

Commit:
video: fbdev: starfive: fix fb bug about HDMI display

mcd500 commented:
I rebased the patch on top of starlight branch after contents have changed from 5.12.y to 5.13.y.
This patch is to be meant for upstream by adding minimum changes to riscv/defconfig which only with enabling jh71x0 soc.

mcd500 commented:
I will rebase it every time when the contents of the starlight branch rebases to new the kernel version.
The original PR was https://github.com/esmil/linux/pull/2

geertu committed:
video: fbdev: starfive: FB_STARFIVE_HDMI_TDA998X depends on DRM_FBDEV_EMULATION

riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `.L0 ':
tda998x.c:(.text+0x51c): undefined reference to `drm_encoder_cleanup'
riscv64-linux-gnu-ld: tda998x.c:(.text+0x534): undefined reference to `drm_encoder_cleanup'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `.L75':
tda998x.c:(.text+0x564): undefined reference to `drm_of_find_possible_crtcs'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `tda998x_encoder_destroy':
tda998x.c:(.text+0x58a): undefined reference to `drm_encoder_init'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `tda998x_bind':
tda998x.c:(.text+0x5b2): undefined reference to `drm_bridge_attach'
riscv64-linux-gnu-ld: tda998x.c:(.text+0x5c0): undefined reference to `drm_encoder_cleanup'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `.L0 ':
tda998x.c:(.text+0x692): undefined reference to `drm_bridge_remove'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `.L124':
tda998x.c:(.text+0x904): undefined reference to `hdmi_infoframe_pack'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `.L135':
tda998x.c:(.text+0x9e8): undefined reference to `drm_connector_cleanup'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `.L0 ':
tda998x.c:(.text+0xa00): undefined reference to `drm_connector_cleanup'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `.L143':
tda998x.c:(.text+0xa56): undefined reference to `drm_connector_init'
riscv64-linux-gnu-ld: tda998x.c:(.text+0xa66): undefined reference to `drm_connector_attach_encoder'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `tda998x_bridge_detach':
tda998x.c:(.text+0xa8a): undefined reference to `__drm_err'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `.L189':
tda998x.c:(.text+0xd16): undefined reference to `drm_do_get_edid'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `.L0 ':
tda998x.c:(.text+0xd34): undefined reference to `drm_connector_update_edid_property'
riscv64-linux-gnu-ld: tda998x.c:(.text+0xd4e): undefined reference to `drm_add_edid_modes'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `.L165':
tda998x.c:(.text+0xd5a): undefined reference to `drm_detect_monitor_audio'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `.L0 ':
tda998x.c:(.text+0xfb0): undefined reference to `__drm_dbg'
riscv64-linux-gnu-ld: tda998x.c:(.text+0x108a): undefined reference to `drm_kms_helper_hotplug_event'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `.L283':
tda998x.c:(.text+0x1844): undefined reference to `drm_hdmi_avi_infoframe_from_display_mode'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `.L361':
tda998x.c:(.text+0x2078): undefined reference to `drm_bridge_add'
riscv64-linux-gnu-ld: drivers/video/fbdev/starfive/tda998x.o: in function `.LANCHOR0':
tda998x.c:(.rodata+0x90): undefined reference to `drm_helper_connector_dpms'
riscv64-linux-gnu-ld: tda998x.c:(.rodata+0x98): undefined reference to `drm_atomic_helper_connector_reset'
riscv64-linux-gnu-ld: tda998x.c:(.rodata+0xb0): undefined reference to `drm_helper_probe_single_connector_modes'
riscv64-linux-gnu-ld: tda998x.c:(.rodata+0xd8): undefined reference to `drm_atomic_helper_connector_duplicate_state'
riscv64-linux-gnu-ld: tda998x.c:(.rodata+0xe0): undefined reference to `drm_atomic_helper_connector_destroy_state'

Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>

geertu committed:
video: fbdev: starfive: Fix frame buffer reserved memory resource conflict

    starfive,vpp-lcdc 12000000.sfivefb: can't request region for resource [mem 0xfb000000-0xfcffffff]
    starfive,vpp-lcdc 12000000.sfivefb: Fail to allocate video RAM
    starfive,vpp-lcdc 12000000.sfivefb: starfive fb init fail
    starfive,vpp-lcdc 12000000.sfivefb: fb info init FAIL
    starfive,vpp-lcdc: probe of 12000000.sfivefb failed with error -16

devm_ioremap_resource() calls devm_request_mem_region(), which fails as
the reserved memory for the frame buffer is already present in the
resource list, cfr. /proc/iomem:

    fb000000-fcffffff : Reserved

Fix this by mapping the frame buffer memory using devm_ioremap().

Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>

mcd500 commented:
Main reason I started to make the minimum changes on defconfig was for adding jh7100 enablement for the upstream defconfig. I remember the upstream maintainer hated to have different defconfig file for each different board when new products show up in the market and end up having huge number of files under arch/arm/configs, so having arch/riscv/configs/defconfig only touching hw enablement.

stffrdhrn commented:
Hi @mcd500 -san, Can you use `starlight_defconfig`? As mentioned in the readme?  https://github.com/esmil/linux/blob/starlight/README.md

stffrdhrn commented:
> Hi @mcd500 -san, Can you use `starlight_defconfig`? As mentioned in the readme? https://github.com/esmil/linux/blob/starlight/README.md

Nevermind, I see you said the upstream maintainer just wants one defconfig.  I don't agree with that so much, but ok.

teknoraver committed:
net: stmmac: use GFP_DMA32

Signed-off-by: Matteo Croce <mcroce@microsoft.com>

geertu committed:
sifive/sifive_l2_cache: Print a backtrace on out-of-range flushes

This makes it easier to find out which driver passes a wrong address
range.

Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>

mcd500 commented:
@stffrdhrn This is the discussion in the past. 

Akira Tsukamoto:  10 days ago
```
@Esmil I tried both your starlight_defconfig and my additions on the original defconfig on the fedora rootfs and perfectly working fine. No Ethernet link down, hartbeat LED is blinking, reboot is working fine. Only difference is loading extra firmwares from your starlight_defconfig is not in my defconfig. The size of kernels from starlight_defconfig is about 1MiB smaller than revised defconfig so starlight_defconfig is definitely the first choice for the BeagleV developer’s.
```
Esmil:  10 days ago
```
That's good to hear. Thank you for testing it!
```

mcd500 commented:
@stffrdhrn The upstream kernel under arch/arm64/configs end up having only one file, defconfig. So what can we do...

mcd500 commented:
Rebased is on top of starlight after moving to 5.13.rc3.

geertu committed:
[WIP] dt-bindings: dma: dw-axi-dmac: Increase DMA channel limit to 16

The first DMAC instance in the StarFive JH7100 SoC supports 16 DMA
channels.

FIXME Given there are more changes to the driver than just increasing
      DMAC_MAX_CHANNELS, we probably need a new compatible value, too.

Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>

pdp7 committed:
dt-bindings: gpio: add starfive,jh7100-gpio bindings

Add bindings for the GPIO controller in the StarFive JH7100 SoC [1].

[1] https://github.com/starfive-tech/beaglev_doc

Signed-off-by: Drew Fustini <drew@beagleboard.org>
Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>

geertu committed:
[WIP] clk: starfive: Add preliminary JH7100 Clock Generator Driver

Add a preliminary driver for the StarFive JH7100 Clock Generator,
based on work by Ahmad Fatoum for Barebox.

Functional differences compared to the Barebox driver:
  - Addition of starfive_clk_pll_mult(), to make the PLL outputs
    reasonable,
  - Add temporary overrides for critical clocks (uart, i2c, spi) until
    we get the clock tree right, so we don't have to keep dummy clocks
    like uartclk and hs_uartclk in the DTS.

This driver sets clk_ignore_unused, else the system locks up when all
clocks deemed unused are disabled.

To be updated when the documentation becomes available.

Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>

geertu committed:
[WIP] dt-bindings: clock: starfive: Add preliminary JH7100 bindings

Add preliminary Device Tree bindings for the StarFive JH7100 Clock
Generator.

To be verified against documentation when it becomes available.

Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>

esmil committed:
dt-bindings: hwmon: add starfive,jh7100-temp bindings

Add bindings for the temperature sensor on the StarFive JH7100 SoC.

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>
Reviewed-by: Rob Herring <robh@kernel.org>

esmil committed:
hwmon: (sfctemp) Add StarFive JH7100 temperature sensor

Register definitions and conversion constants based on sfctemp driver by
Samin in the StarFive 5.10 kernel.

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>
Signed-off-by: Samin Guo <samin.guo@starfivetech.com>

atishp04 committed:
RISC-V: Support non-coherent DMA operations

** Do not upstream **

This is hacky fix just for testing. The actual patch would read the
RISCV_UNCACHED_OFFSET from the DT for only the non-coherent devices.
All other devices on beagleV and all other platform should just set
dma_default_coherent to true.

[Emil: remove spurious whitespace and fix format string warning]

Signed-off-by: Atish Patra <atish.patra@wdc.com>
Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>

atishp04 committed:
sifive/sifive_l2_cache: Align the address to cache line

[Emil: fix suggested by Geert Uytterhoeven <geert@linux-m68k.org>]

Signed-off-by: Atish Patra <atish.patra@wdc.com>
Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>

geertu committed:
video: fbdev: starfive: Use round_up() instead of _ALIGN_UP()

_ALIGN_UP() is being removed by commit efcec32fe84ab4ee ("riscv: Cleanup
unused functions") in riscv/for-next:

    drivers/video/fbdev/starfive/starfive_displayer.c: In function ‘of_parse_wr_cmd’:
    drivers/video/fbdev/starfive/starfive_displayer.c:616:27: error: implicit declaration of function ‘_ALIGN_UP’ [-Werror=implic
    it-function-declaration]
      616 |  dev_cmds->cmds = kzalloc(_ALIGN_UP(alloc_bytes, 4), GFP_KERNEL);
	  |                           ^~~~~~~~~

Just use roundup() instead.

Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>
Tested-by: Emil Renner Berthing <kernel@esmil.dk>

esmil committed:
dt-bindings: mfd: convert tps65086.txt to YAML

This converts the tps65086.txt binding description to YAML schema so
dts files can be verified automatically.

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>

teknoraver committed:
riscv: use the generic string routines

Use the generic routines which handle alignment properly.

Signed-off-by: Matteo Croce <mcroce@microsoft.com>
Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>

geertu committed:
[WIP] dt-bindings: clock: starfive: Add JH7100 Clock Definitions

Add all clock outputs for the StarFive JH7100 Clock Generator.

Based on work by Ahmad Fatoum for Barebox, with "JH7100_" prefixes added
to all definitions.

Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>

geertu committed:
[WIP] dt-bindings: reset: starfive: Add JH7100 Reset Definitions

Add all resets for the StarFive JH7100 Reset Controller.

Based on work by Ahmad Fatoum for Barebox, with "JH7100_" prefixes added
to all definitions.

Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>

geertu committed:
[NOT-FOR-UPSTREAM] clk: Make clk_ignore_unused public

Make clk_ignore_unused public, so the StarFive JH7100 Clock Generator
driver can set the flag.

Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>

geertu committed:
drivers/dma: dw-axi-dmac-starfive: Remove calls specific to ARM64 ACPI

iort_dma_setup() is being removed by commit db59e1b6e49201be ("ACPI:
arm64: Move DMA setup operations out of IORT") in iommu/next:

    drivers/dma/dw-axi-dmac-starfive/starfive_dmaengine_memcpy.c: In function ‘dw_dma_async_
    do_memcpy’:
    drivers/dma/dw-axi-dmac-starfive/starfive_dmaengine_memcpy.c:152:2: error: implicit decl
    aration of function ‘iort_dma_setup’ [-Werror=implicit-function-declaration]
      152 |  iort_dma_setup(dma_dev, &dma_addr, &dma_size);
	  |  ^~~~~~~~~~~~~~
    drivers/dma/dw-axi-dmac-starfive/starfive_dmaengine_memcpy.c:153:8: warning: assignment to ‘const struct iommu_ops *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
      153 |  iommu = iort_iommu_configure_id(dma_dev, NULL);
	  |        ^
    drivers/dma/dw-axi-dmac-starfive/starfive_dmaengine_memcpy.c: In function ‘dw_dma_memcpy_raw’:
    drivers/dma/dw-axi-dmac-starfive/starfive_dmaengine_memcpy.c:223:8: warning: assignment to ‘const struct iommu_ops *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
      223 |  iommu = iort_iommu_configure_id(dma_dev, NULL);
	  |        ^

iort_dma_setup() and iort_iommu_configure_id() are part of the ARM64
ACPI implementation.  As CONFIG_ACPI_IORT cannot be enabled on RISC-V,
they were dummies anyway, so these calls can just be removed.

[Emil: remove unused local variables too]

Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>
Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>
---
Boot-tested, but the affected code paths were not exercised.

pdp7 committed:
gpio: starfive-jh7100: Add StarFive JH7100 GPIO driver

Add GPIO driver for the StarFive JH7100 SoC [1] used on the
BeagleV Starlight JH7100 board [2].

[1] https://github.com/starfive-tech/beaglev_doc/
[2] https://github.com/beagleboard/beaglev-starlight

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>
Signed-off-by: Huan Feng <huan.feng@starfivetech.com>
Signed-off-by: Drew Fustini <drew@beagleboard.org>

Commit:
[NOT-FOR-UPSTREAM] riscv: Add starfive jh7100 starlight fedora defconfig

Signed-off-by: TekkamanV <tekkamanv@starfivetech.com>

Commit:
media: starfive: Remove MODULE_SUPPORTED_DEVICE macro call

This macro was never implemented and completely removed upstream.

Signed-off-by: TekkamanV <tekkamanv@starfivetech.com>

pdp7 committed:
dt-bindings: riscv: add starfive jh7100 bindings

Add DT binding documentation for the StarFive JH7100 Soc [1] and the
BeagleV Starlight JH7100 board [2].

[1] https://github.com/starfive-tech/beaglev_doc
[2] https://github.com/beagleboard/beaglev-starlight

Signed-off-by: Drew Fustini <drew@beagleboard.org>

teknoraver committed:
lib/string: optimized memcpy

Patch series "lib/string: optimized mem* functions", v2.

Rewrite the generic mem{cpy,move,set} so that memory is accessed with the
widest size possible, but without doing unaligned accesses.

This was originally posted as C string functions for RISC-V[1], but as
there was no specific RISC-V code, it was proposed for the generic
lib/string.c implementation.

Tested on RISC-V and on x86_64 by undefining __HAVE_ARCH_MEM{CPY,SET,MOVE}
and HAVE_EFFICIENT_UNALIGNED_ACCESS.

These are the performances of memcpy() and memset() of a RISC-V machine on
a 32 mbyte buffer:

memcpy:
original aligned:	 75 Mb/s
original unaligned:	 75 Mb/s
new aligned:		114 Mb/s
new unaligned:		107 Mb/s

memset:
original aligned:	140 Mb/s
original unaligned:	140 Mb/s
new aligned:		241 Mb/s
new unaligned:		241 Mb/s

The size increase is negligible:

$ scripts/bloat-o-meter vmlinux.orig vmlinux
add/remove: 0/0 grow/shrink: 4/1 up/down: 427/-6 (421)
Function                                     old     new   delta
memcpy                                        29     351    +322
memset                                        29     117     +88
strlcat                                       68      78     +10
strlcpy                                       50      57      +7
memmove                                       56      50      -6
Total: Before=8556964, After=8557385, chg +0.00%

These functions will be used for RISC-V initially.

[1] https://lore.kernel.org/linux-riscv/20210617152754.17960-1-mcroce@linux.microsoft.com/

This patch (of 3):

Rewrite the generic memcpy() to copy a word at time, without generating
unaligned accesses.

The procedure is made of three steps: First copy data one byte at time
until the destination buffer is aligned to a long boundary.  Then copy the
data one long at time shifting the current and the next long to compose a
long at every cycle.  Finally, copy the remainder one byte at time.

This is the improvement on RISC-V:

original aligned:	 75 Mb/s
original unaligned:	 75 Mb/s
new aligned:		114 Mb/s
new unaligned:		107 Mb/s

and this the binary size increase according to bloat-o-meter:

Function     old     new   delta
memcpy        36     324    +288

Link: https://lkml.kernel.org/r/20210702123153.14093-1-mcroce@linux.microsoft.com
Link: https://lkml.kernel.org/r/20210702123153.14093-2-mcroce@linux.microsoft.com
Signed-off-by: Matteo Croce <mcroce@microsoft.com>
Cc: Nick Kossifidis <mick@ics.forth.gr>
Cc: Guo Ren <guoren@kernel.org>
Cc: Christoph Hellwig <hch@infradead.org>
Cc: David Laight <David.Laight@aculab.com>
Cc: Palmer Dabbelt <palmer@dabbelt.com>
Cc: Emil Renner Berthing <kernel@esmil.dk>
Cc: Drew Fustini <drew@beagleboard.org>
Cc: Nick Desaulniers <ndesaulniers@google.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Stephen Rothwell <sfr@canb.auug.org.au>

teknoraver committed:
lib/string: optimized memmove

When the destination buffer is before the source one, or when the buffers
doesn't overlap, it's safe to use memcpy() instead, which is optimized to
use a bigger data size possible.

This "optimization" only covers a common case.  In future, proper code
which does the same thing as memcpy() does but backwards can be done.

Link: https://lkml.kernel.org/r/20210702123153.14093-3-mcroce@linux.microsoft.com
Signed-off-by: Matteo Croce <mcroce@microsoft.com>
Cc: Christoph Hellwig <hch@infradead.org>
Cc: David Laight <David.Laight@aculab.com>
Cc: Drew Fustini <drew@beagleboard.org>
Cc: Emil Renner Berthing <kernel@esmil.dk>
Cc: Guo Ren <guoren@kernel.org>
Cc: Nick Desaulniers <ndesaulniers@google.com>
Cc: Nick Kossifidis <mick@ics.forth.gr>
Cc: Palmer Dabbelt <palmer@dabbelt.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Stephen Rothwell <sfr@canb.auug.org.au>

teknoraver committed:
lib/string: optimized memset

The generic memset is defined as a byte at time write.  This is always
safe, but it's slower than a 4 byte or even 8 byte write.

Write a generic memset which fills the data one byte at time until the
destination is aligned, then fills using the largest size allowed, and
finally fills the remaining data one byte at time.

On a RISC-V machine the speed goes from 140 Mb/s to 241 Mb/s, and this the
binary size increase according to bloat-o-meter:

Function     old     new   delta
memset        32     148    +116

Link: https://lkml.kernel.org/r/20210702123153.14093-4-mcroce@linux.microsoft.com
Signed-off-by: Matteo Croce <mcroce@microsoft.com>
Cc: Christoph Hellwig <hch@infradead.org>
Cc: David Laight <David.Laight@aculab.com>
Cc: Drew Fustini <drew@beagleboard.org>
Cc: Emil Renner Berthing <kernel@esmil.dk>
Cc: Guo Ren <guoren@kernel.org>
Cc: Nick Desaulniers <ndesaulniers@google.com>
Cc: Nick Kossifidis <mick@ics.forth.gr>
Cc: Palmer Dabbelt <palmer@dabbelt.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Stephen Rothwell <sfr@canb.auug.org.au>
